### ISSUE TITLE
ExecuteListener's fetchEnd method receives no Query for DML queries

### ISSUE THREAD

**Issue by fcv** [jOOQ/jOOQ#15061]
### Expected behavior

As per  [`fetchEnd`'s Javadoc](https://www.jooq.org/javadoc/3.17.12/org.jooq/org/jooq/ExecuteListener.html#fetchStart(org.jooq.ExecuteContext)) its `ExecuteContext`'s `query()` method returns:

> The Query object, if a jOOQ query is being executed or null otherwise

So, for a code like: 

```java
        ResultQuery<?> resultQuery = ctx.update(TEST).set(TEST.CD, 1).returning(TEST.CD);
        resultQuery.fetch();
```

I would expect the `query()` method of the `ExecuteContext` received by all registered `ExecuteListener` instances to return the `resultQuery` instance.

### Actual behavior

It's been observed that for some types of Queries it indeed works like this. However apparently for DML queries with RETURNING clauses  `ExecuteContext.query()` return `null`.

A reproducible case using jOOQ version 3.18.2 can be found in https://github.com/fcv/jOOQ-mcve, more specifically in [test/java/postgres/JavaTest](https://github.com/fcv/jOOQ-mcve/blob/53c73009cf88e24e414c37543d741f1b44e0c0fd/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/JavaTest.java#L146-L153).

fwiw, I could also reproduce this behaviour in older version, like 3.11.x.

I'm not sure whether it's related, or even expected, but for some reason the `ExecuteContext` instance received by `fetchEnd` method is [a completely different instance](https://github.com/jOOQ/jOOQ/blob/version-3.18.2/jOOQ/src/main/java/org/jooq/impl/AbstractDMLQuery.java#L1228-L1229) from the one received by, for example, `executedEnd`.

Originally reported in https://groups.google.com/g/jooq-user/c/CnQZxmcXfHc.

### Steps to reproduce the problem

https://github.com/fcv/jOOQ-mcve/blob/53c73009cf88e24e414c37543d741f1b44e0c0fd/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/JavaTest.java

### jOOQ Version

3.18.2

### Database product and version

PostgreSQL 15.3 

### Java Version

openjdk 17.0.4 2022-07-19 LTS

### OS Version

macOS Ventura 13.2.1

**Comment by fcv** [comment]
I'm not sure whether it's related or not, but I noticed another strange behaviour related to execution of DML with RETURNING clause. Apparently ExecuteListener's `end` method is executed twice.

I have added a failing test case in the same https://github.com/fcv/jOOQ-mcve repository, in a different files [ExecuteListenerExecCountTest](https://github.com/fcv/jOOQ-mcve/blob/execute-listener-exec-count-test/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/ExecuteListenerExecCountTest.java). In this class, test method [`plainUpdateWithReturningClause`](https://github.com/fcv/jOOQ-mcve/blob/288d0b2ab5f67efc00d6e94507de0104eb817d5e/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/ExecuteListenerExecCountTest.java#L159-L167) fails with a report like:

```
Failed tests:   dslUpdateWithReturningClause(org.jooq.mcve.test.java.postgres.ExecuteListenerExecCountTest): counterExecuteListener.end should be executed exactly once expected:<1> but was:<2>
```

**Comment by lukaseder** [comment]
Thanks a lot for your report and for your reproducer. I'll try to reproduce this on my side.

**Comment by lukaseder** [comment]
Might be a duplicate of https://github.com/jOOQ/jOOQ/issues/14306

**Comment by lukaseder** [comment]
Yes, definitely a duplicate of #14306. I'm not sure if there's a simple workaround for the time being. There are currently 2 `ExecuteContext` instances and lifecycles, which don't communicate with each other...

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/14306 has been implemented in jOOQ 3.19.0. As this may break assumptions of users who have started to rely on 2 `ExecuteContext` lifecycles to be created, I won't backport this change to earlier releases.
